### ISSUE TITLE
cell data not recorded

### ISSUE THREAD

**Issue by matt755** [mjaakko/NeoStumbler#360]
### Description

Neostumbler detects and records wi-fi and Bluetooth beacons, but no cell Tower information is recorded.
A parallel installation of the app Tower Collector does record cells, so I assume an issue with neostumbler. 

### App variant

full

### Version

1.4.1

### Device

OnePlus ONE A2003

### Android version

10

### New bug

- [X] I have checked that this bug has not been reported before

**Comment by mjaakko** [comment]
Are you using a custom ROM?

**Comment by matt755** [comment]
Yes, it's a LineageOS based build from the e. foundation (e_oneplus2).
The build is not available anymore and android 10 is eol anyway. So don't invest time if you cannot think of a quick fix.

**Comment by mjaakko** [comment]
Alright. NeoStumbler is using some of the new APIs introduced in Android 10 for cell scanning and I'm guessing those might not be properly supported on custom ROMs.

As a workaround, you could try revoking the phone permission from NeoStumbler. Without the phone permission, cell scanning is only done on the primary SIM card, which doesn't use as many new APIs. Not sure if it works, but it's worth trying

A proper fix would be adding support to use legacy APIs for cell scanning, but yeah I don't think it's worth the effort given that this should only affect a small number of old devices

**Comment by skinkie** [comment]
I also noticed that Towercollector collects much more cell information. I am not running a custom rom, but I wanted to validate that this is not caused by some permission issue. Hence trying trouble shooting and fix the **first issue**. I wonder why those persmissions were not asked for or granted on two independent installations.

**Comment by MatthiasHGW** [comment]
I do have the same problem with the current version from F-droid running on LineageOS 17.1 (the last version available for my old phone) not collecting cell tower information. Revoking phone permission did not help in my case. 

I just noticed that LineageOS 17 is based on Android 10, so I wonder if an old API is the underlying issue. Is there an easy way to check, can I somehow collect debugging messages?

**Comment by mjaakko** [comment]
@MatthiasHGW the release build does not have logging enabled. You would have to build a debug version and then use logcat to see the logs. There's logging for failed cell scans: https://github.com/mjaakko/NeoStumbler/blob/c3fdf62dd704fb04bf5478f83c46f247fbee2ab3/app/src/main/java/xyz/malkki/neostumbler/scanner/source/TelephonyManagerCellInfoSource.kt#L60 - if that message never appears and no cell data os collected, then there's either an issue in the data which causes it to be filtered (no cell ID etc.) or the cell scan callback is never invoked

**Comment by matt755** [comment]
I think I found out the reason for the failed recording of cell tower data. For whatever reason in my case the mobileNetworkCode cannot be identified (mMnc=null):

`CellInfoLte:{mRegistered=YES mTimeStamp=440378202701337ns mCellConnectionStatus=2147483647 CellIdentityLte:{ mCi=20667139 mPci=445 mTac=48023 mEarfcn=14 mBandwidth=2147483647 mMcc=262 mMnc=null mAlphaLong= mAlphaShort=} CellSignalStrengthLte: rssi=2147483647 rsrp=2147483647 rsrq=2147483647 rssnr=2147483647 cqi=2147483647 ta=0 level=0 android.telephony.CellConfigLte :{ isEndcAvailable = false }}`

Replacing the function "fillMissingData" in TelephonyManagerCellInfoSource for test purposes with the following code fixed the issue.
So, if the mobileNetworkCode is not mandatory for collecting data, and if the root cause cannot be identified I suggest to make it optional, i.e. fill in only if available, otherwise set a default or leave empty, if the api allows this.

```
    private fun List<CellTower>.fillMissingData(): List<CellTower> {
        if (size == 1) {
            return this
        } else {
            val mobileCountryCodes = mapNotNull { it.mobileCountryCode }.toSet()
            var mobileNetworkCodes = mapNotNull { it.mobileNetworkCode }.toSet()

            if (mobileNetworkCodes.isEmpty()) {
                mobileNetworkCodes = setOf("00")
            } else if (mobileNetworkCodes.size > 1) {
                val firstMobileNetworkCode = mobileNetworkCodes.firstOrNull() ?: "00"
                mobileNetworkCodes = setOf(firstMobileNetworkCode)
            }

            return map { cellTower ->
                cellTower.copy(
                    mobileCountryCode = cellTower.mobileCountryCode ?: mobileCountryCodes.first(),
                    mobileNetworkCode = cellTower.mobileNetworkCode ?: mobileNetworkCodes.first()
                )
            }
        }
    }
```

P.S. Tower Collector uses the implementation https://github.com/mroczis/netmonster-core for evaluation of the cell data and correctly identifies the mnc.

**Comment by mjaakko** [comment]
MNC is needed to uniquely identify the cell tower. Cell IDs are not globally unique.

I'll take a look at the library you've linked. It could make cell tower scanning more robust

**Comment by mjaakko** [comment]
In the latest rc (1.5.0-rc2), there is a possible workaround for this. Try if it works for you

**Comment by matt755** [comment]
Works like a charm!
